### PR TITLE
Separated student and professor lecture components

### PR DESCRIPTION
**Pull request by loganallen**
- Different student and lecture components
- Basic functionality for posting a question, students receive and select an option, then response is sent back to server upon professor ending the question
- Initial PR so we can separate client/server repos (needs fixing up later and better server-side data structures)

### CODE REVIEW

**Comment by Jma353** [comment]
@dantheli @yassinmziya plz review

File: client/src/components/common/ClickerQuestion.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    };
+  }
+
+  render(): void {
+    const choices = this.state.choices.map((el, i) => (
+      <ClickerButton
+        key={i}
+        onClick={ () => this.props.onClick(i) }
+        inverse={(i == this.props.selection)}
+      >
```
Comment (Jma353) [inline client/src/components/common/ClickerQuestion.js]: Move this up, hanging `>` looks weird.

File: client/src/components/common/ClickerQuestion.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  onClick: number => void,
+  selection: number
+};
+type State = {
+  question: string,
+  choices: Object
+}
+
+// TODO - Make this actually nice
+class ClickerQuestion extends React.Component<Props, State> {
```
Comment (Jma353) [inline client/src/components/common/ClickerQuestion.js]: under this:

```javascript
props: Props;
state: State;
```

Comment (Jma353) [reply client/src/components/common/ClickerQuestion.js]: This is something we should do everywhere btw - since these are both instance variables

Comment (loganallen) [reply client/src/components/common/ClickerQuestion.js]: does having `React.Component<Props, State>` not define the respective prop and state types?

File: client/src/components/lecture/LectureStudent.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        <ClickerTextInput
+          placeholder='Enter lecture id'
+          value={this.state.lectureId}
+          onChange={this._onLectureIdChange}
+        />
+        <ClickerButton
+          onClick={ () => this._onJoinLecture(this.state.lectureId) }
+          inverse={false}
+          disabled={!this.state.lectureId}
+        >
```
Comment (Jma353) [inline client/src/components/lecture/LectureStudent.js]: ditto

File: client/src/components/lecture/LectureStudent.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  selection: number
+};
+
+class LectureStudent extends React.Component<Props, State> {
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      socket: undefined,
```
Comment (Jma353) [inline client/src/components/lecture/LectureStudent.js]: `null`, never `undefined`

File: client/src/components/lecture/LectureStudent.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+import ClickerPage from '../common/ClickerPage';
+import ClickerTextInput from '../common/ClickerTextInput';
+import ClickerButton from '../common/ClickerButton';
+import ClickerQuestion from '../common/ClickerQuestion';
+
+type Props = {
+};
+type State = {
+  socket: Socket,
```
Comment (Jma353) [inline client/src/components/lecture/LectureStudent.js]: `?Socket` if it has the possibility of being `null`

File: client/src/components/lecture/LectureStudent.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  connected: boolean,
+  lectureId: string,
+  currentLecture: string,
+  question: Object,
```
Comment (Jma353) [inline client/src/components/lecture/LectureStudent.js]: Is this something we might change to be a concrete type in the future, given that a `Question` is going to appear and reappear with expected fields?  Just a thought - possibly `// TODO` this?

File: client/src/components/lecture/LectureStudent.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+      this.setState({
+        question: {},
+        selection: -1
+      });
+    });
+  }
+
+  // Close socket on component dismount
+  componentWillUnmount(): void {
+    this.state.socket.close();
```
Comment (Jma353) [inline client/src/components/lecture/LectureStudent.js]: `this.state.socket && this.state.socket.close()`, as `this.state.socket` can be null.

File: client/src/components/common/ClickerQuestion.js (hunk 2 above)
Comment (Jma353) [reply client/src/components/common/ClickerQuestion.js]: I'd always declare an instance variable `state`, and habitually declare a `props` one too.

**Comment by Jma353** [comment]
This is a lot of quality work!  But this PR is kinda huge - we should try and make them smaller and more digestible in the future, although I know this was a big change.  In addition, there's a lot of data-types and functionality that is currently in a `// TODO` state due to pending designs - we should make all these with `// TODO`'s as such so we can easily search for them in the future and add the functionality, Flow types, etc. in their proper places.

File: client/src/components/lecture/LectureStudent.js (hunk 6 above)
Comment (loganallen) [reply client/src/components/lecture/LectureStudent.js]: Yeah I had the same thought, but just made this arbitrary `Object` for now. Will have something like `type Question` later on

**Comment by loganallen** [comment]
Agreed--got carried away with different things (like making question components) but should have done that separately. Will def have smaller tasks/PRs in future!

**Pull request merged**